plugins/custom-auth/index.js

```diff
@@ -0,0 +1,113 @@
+const Users = require('services/users');
+const debug = require('debug')('talk-plugin-custom-auth');
+const authz = require('middleware/authorization');
+const { get } = require('lodash');
+
+module.exports = {
+
+  // When the JWT is passed to Talk, first it will be validated
+  // Next, Talk will attempt to locate a user in its DB using the token's sub claim,
+  // if the user is not found, the tokenUserNotFound hook will be called to create the user
+  tokenUserNotFound: async ({ jwt }) => {
+    debug('The jwt is ' + JSON.stringify(jwt));
+    
+    // Since the JWT has already been validated we can pass it's claims directly to upsertExternalUser
+    const user = await Users.upsertExternalUser(
+      null,
+      jwt.sub,
+      jwt.iss,
+      jwt.username
+    );
+
+    // Persisting email address in Talk is required only if sending notifications from Talk.
+    // If email was included on the JWT we can add it to a "local" profile
+    // and push that into the user's Profiles array.
+    // upsertExternalUser above will also create a profile as: { provider:jwt.iss, id:jwt.sub }
+    const email = jwt.email.toLowerCase();
```

Comment:
To guard against double adding, you should check `if (user.wasUpserted) { }` before pushing/saving the profile.

Reply:
I'll add this as a comment